kratos/utilities/plane_approximation_utility.h

```diff
@@ -0,0 +1,244 @@
+    ///@{
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief This function computes the plane base point
+     * @param rPointsCoords Vector containing the set of point coordinates
+     * @return rBasePointCoords Plane base point coordinates
+     */
+    static void GetPlaneBasePoint(
+        const std::vector< array_1d<double,3> > &rPointsCoords,
+        array_1d<double,3> &rBasePointCoords) 
+    {
+        const unsigned int n_points = rPointsCoords.size();
+        KRATOS_ERROR_IF(n_points == 0) << "No base point can be computed. Points container is empty." << std::endl;
+
+        noalias(rBasePointCoords) = ZeroVector(3);
+        for (unsigned int j = 0; j < n_points; ++j){
+            rBasePointCoords += rPointsCoords[j];
+        }
+        rBasePointCoords /= n_points;
+    }
+
+    /**
+     * @brief This function sets the matrix A as trans(P)*P
+     * @param rPointsCoords Vector containing the set of point coordinates
+     * @param rBasePointCoords Plane base point coordinates
+     * @return rA A matrix
+     */
+    static void SetMatrixA(
+        const std::vector< array_1d< double,3 > > &rPointsCoords,
+        const array_1d<double,3> &rPlaneBasePointCoords,
+        bounded_matrix<double,3,3> &rA) 
+    {
+        noalias(rA) = ZeroMatrix(3);
+        const unsigned int n_points = rPointsCoords.size();
+
+        for (unsigned int i = 0; i < 3; ++i){
+            const double base_i = rPlaneBasePointCoords(i);
+            for (unsigned int j = 0; j < 3; ++j){
+                const double base_j = rPlaneBasePointCoords(j);
+                for (unsigned int m = 0; m < n_points; ++m){
+                    const double pt_m_i = rPointsCoords[m](i);
+                    const double pt_m_j = rPointsCoords[m](j);
+                    rA(i,j) += (base_i - pt_m_i)*(base_j - pt_m_j);
+                }
+            }
+        }
+    }
+
+    /**
+     * @brief This function computes using the inverse power method the minimal eigenvalue
+     * @param rPointsCoords Vector containing the set of point coordinates
+     * @param rBasePointCoords Plane base point coordinates
+     * @return rPlaneNormal The plane unit normal
+     */
+    static void GetPlaneNormal(
+        const std::vector< array_1d<double,3> > &rPointsCoords,
+        const array_1d<double,3> &rPlaneBasePointCoords,
+        array_1d<double,3> &rPlaneNormal) 
+    {
+        // Solve the A matrix eigenvalue problem 
+        bounded_matrix<double, 3, 3> a_mat, eigenval_mat, eigenvector_mat;
+        SetMatrixA(rPointsCoords, rPlaneBasePointCoords, a_mat);
+        bool converged = MathUtils<double>::EigenSystem<3>(a_mat, eigenvector_mat, eigenval_mat);
+        KRATOS_ERROR_IF(!converged) << "Plane normal can't be computed. Eigenvalue problem did not converge." << std::endl;
+
+        // Find the minimum eigenvalue
+        double min_eigval = 0.0;
```

Comment:
Are you sure that min_eigval is always negative? It shouldn't be initialized with std::numerical_limits::max?

Reply:
You're right, I'm changing it.

Reply:
Indeed, this yields an error in 2D cases... Thanks a lot @pooyan-dadvand .